src/custom_links/singly_linked_list.rs

```diff
@@ -0,0 +1,1466 @@
+// Copyright 2020 Amari Robinson
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+use core::fmt;
+use core::cell::Cell;
+
+use super::Adapter;
+
+use crate::IntrusivePointer;
+
+// =============================================================================
+// Link
+// =============================================================================
+
+/// Intrusive link that allows an object to be inserted into a `SinglyLinkedList`.
+pub trait Link<NodeRef> {
```

Comment:
This should be an `unsafe trait` (since incorrectly implementing it can cause UB), and all methods should also be unsafe (since you can cause UB by calling them incorrectly).

Reply:
Including `is_linked`? `crate::singly_linked_list::Link::is_linked` is not unsafe.

Reply:
Ah sorry, you're right. Only `set_next` and `force_unlink` need to be unsafe.